pageserver/src/tenant.rs

```diff
@@ -748,14 +748,16 @@ impl Tenant {
                         },
                     )
                 } else {
+                    // SpawnMode::{Create,Eager} always cause jumping ahead of the
+                    // concurrent_tenant_warmup queue
                     AttachType::Normal
                 };
 
                 let preload = match (&mode, &remote_storage) {
```

Comment:
I find it a bit odd that for `preload` we check the spawn mode inline to determine if it's required, but for attach it's done in `Timeline::attach`. I think this would read better if we did the same for preload and moved the timer initalisation in the respective functions (attach and preload) since we match on the spawn mode already.

In any case, this comment doesn't really concern your PR, so feel free to punt.

Reply:
Hmm ok, I did not notice that. I agree looks quite sketchy, but I am bit unclear on what are you requesting.

You suggest:
- Timers on line R761 moved to preload (makes sense)
- Timer on R774 is a bit bogus "histogram used as a counter" but I left it there because I found it there in #6621
- Timer on R832 moved to attach

I'd also love to remove the `make_broken` and use normal `?` for that purpose, remove the extra indentation etc.

---

I do not follow this part:

> I find it a bit odd that for preload we check the spawn mode inline to determine if it's required, but for attach it's done in Timeline::attach.

SpawnMode is matched in this `fn spawn` but the "`TenantPreload::default`" alike functionality is only in `fn attach` for SpawnMode::Create. Do you think that should move up or something else?

Reply:
Though, I am feeling like doing instead another cleanup around `attach` PR (or around `Tenant::spawn` in this case), but would still like to understand your point.

Reply:
> I do not follow this part:

I was trying to suggest changing `Tenant::preload` to return something like `Option<Result<TenantPreload>>`.
Basically, we'd move the matching on `SpawnMode` into `Tenant::preload` and return None if a preload is not required. I think it may decluter `spawn` a bit and make the logic easier to follow, but I don't feel super strongly.

Reply:
> Though, I am feeling like doing instead another cleanup around attach PR (or around Tenant::spawn in this case), but would still like to understand your point.

Doing it in a different PR makes sense